mysql-test/main/partition_innodb.result

```diff
@@ -1046,10 +1046,10 @@ INSERT INTO t1 VALUES ();
 SELECT * FROM t1;
 a
 -1
-1
 3
 4
 6
+7
 DROP TABLE t1;
```

Comment:
Hi @nayuta-yanagisawa !

According to docs https://mariadb.com/kb/en/auto_increment-faq/#manual-notes,
> Inserting a negative number is regarded as inserting a very large positive number

I guess that rule was followed by years, so you can't break it here, especially by pushing in the two years as stable branch. There can be workflows relying on this behavior.

Reply:
Hi, thank you for your comment! 

Sorry for my super misleading comment, but the point is not *negative* integer. Technically speaking, what restores the auto-increment column, in main.partition_innodb, is the unsuccessful insert `INSERT INTO t1 PARTITION(p0) VALUES ()` rather than the following successful inserts.

Please see the following modified version of the test and its result (on the HEAD  of 10.3). The last insert `INSERT INTO t1 VALUES ()` inserts 2 to the column `a`, while the largest value of the auto-increment column is 6. This means that, even though we insert 1 instead of -1, the auto-increment column is restored.

test
```
--echo #
--echo # MDEV-18244 Server crashes in ha_innobase::update_thd / ... / ha_partition::update_next_auto_inc_val
--echo #

CREATE TABLE t1 (a INT)
  ENGINE=InnoDB
  PARTITION BY RANGE (a) (
        PARTITION p0 VALUES LESS THAN (6),
          PARTITION pn VALUES LESS THAN MAXVALUE
      );
INSERT INTO t1 VALUES (4),(5),(6);
ALTER TABLE t1 MODIFY a INT AUTO_INCREMENT PRIMARY KEY;
UPDATE t1 PARTITION (p0) SET a = 3 WHERE a = 5;
--error ER_ROW_DOES_NOT_MATCH_GIVEN_PARTITION_SET
INSERT INTO t1 PARTITION(p0) VALUES ();
INSERT INTO t1 PARTITION(p0) VALUES (1);
INSERT INTO t1 VALUES ();
SELECT * FROM t1;
DROP TABLE t1;

```

result 
```
#
# MDEV-18244 Server crashes in ha_innobase::update_thd / ... / ha_partition::update_next_auto_inc_val
#
CREATE TABLE t1 (a INT)
ENGINE=InnoDB
PARTITION BY RANGE (a) (
PARTITION p0 VALUES LESS THAN (6),
PARTITION pn VALUES LESS THAN MAXVALUE
);
INSERT INTO t1 VALUES (4),(5),(6);
ALTER TABLE t1 MODIFY a INT AUTO_INCREMENT PRIMARY KEY;
UPDATE t1 PARTITION (p0) SET a = 3 WHERE a = 5;
INSERT INTO t1 PARTITION(p0) VALUES ();
ERROR HY000: Found a row not matching the given partition set
INSERT INTO t1 PARTITION(p0) VALUES (1);
INSERT INTO t1 VALUES ();
SELECT * FROM t1;
a
1
2
3
4
6
DROP TABLE t1;
```

Reply:
I still think the correct value is 7, not 1. I also believe that my change does not break the compatibility because inserting a negative value usually does not reset the auto-increment counter. Please see the implementation of `set_auto_increment_if_higher`. https://github.com/MariaDB/server/blob/10.3/sql/ha_partition.h#L1407-L1418

Reply:
Yes, you were, right, sorry. If we remove error-failing insert, then it does produce the same output as in your fix.
Errors should not reset autoincrement either.